File: tools/splc/clock.c
#include "riscv.h"

#define CLINT 0x2000000L
#define MTIME_MMAP_OFFSET 0xBFF8L
#define MTIME_FREQ 4000000L // 4M Hz

static uint64 r_mtime() {
    return r_mem(CLINT + MTIME_MMAP_OFFSET);
}

// delay in microseconds
// 1μs = 1MHz: our CPU reference clock frequency is 4MHz which is used by mtime, so 1 CPU cycle is 1/24 μs
// For 1μs, we need 24 CPU cycles
void delay(uint64 μs)
{
    uint64 cycle1 = r_mtime();
    uint64 cycle2 = cycle1 + μs * (MTIME_FREQ / 1000000);
    do {
        cycle1 = r_mtime();
    } while(cycle2 >= cycle1);
}

File: src/kernel/hal/jh7110/clock.zig
const peripherals = @import("../../soc/jh7110.zig").devices.JH71110.peripherals;

fn mtime() usize {
    return peripherals.sifive_clint0_0.mtime.raw;
}

pub const TimeUnit = enum(u2) {
    sec,
    ms,
    us,
};

const TICKS_PER_SECOND = 4_000_000; //4M Hz;

// delay in seconds/milliseconds/microseconds
// 1μs = 1MHz: our CPU reference clock frequency is TICKS_PER_SECOND Hz which is used by mtime
// so 1 cycle = 4μs
// μs is highly inacurate
pub fn delay(n: usize, unit: TimeUnit) void {
    var ticks: usize = switch (unit) {
        .sec => n * TICKS_PER_SECOND,
        .ms => n * TICKS_PER_SECOND / 1_000,
        .us => n * TICKS_PER_SECOND / 1_000_000,
    };
    const t = mtime() + ticks;
    while (mtime() < t) {}
}
